pkg/webhook/bootstrap.go

```diff
@@ -76,89 +81,125 @@ func (s *Server) setBootstrappingDefault() {
 	if s.Service != nil && s.Port != 443 {
 		s.Port = 443
 	}
-	if s.CertProvisioner == nil {
-		s.CertProvisioner = &cert.Provisioner{
-			Client: s.Client,
-		}
+	if s.certProvisioner == nil {
+		s.certProvisioner = &cert.Provisioner{}
+	}
+	if s.CertGenerator == nil {
+		s.CertGenerator = &generator.SelfSignedCertGenerator{}
+	}
+	if s.Writer == nil {
+		s.Writer = os.Stdout
 	}
 }
 
 // bootstrap returns the configuration of admissionWebhookConfiguration in yaml format if dryrun is true.
 // Otherwise, it creates the the admissionWebhookConfiguration objects and service if any.
 // It also provisions the certificate for your admission server.
-func (s *Server) bootstrap(dryrun bool) ([]byte, error) {
+func (s *Server) bootstrap(dryrun bool) error {
+	// do defaulting if necessary
 	s.once.Do(s.setDefault)
 
 	var err error
 	s.mutatingWebhookConfig, err = s.mutatingWHConfigs()
 	if err != nil {
-		return nil, err
-	}
-	err = s.CertProvisioner.Sync(s.mutatingWebhookConfig)
-	if err != nil {
-		return nil, err
+		return err
 	}
 
 	s.validatingWebhookConfig, err = s.validatingWHConfigs()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	err = s.CertProvisioner.Sync(s.validatingWebhookConfig)
-	if err != nil {
-		return nil, err
+
+	cc, err := s.clientConfig()
```

Comment:
Maybe you can have another function like `getConfigs` to get all the configs needed in bootstrap: `mutatinWebhookConfig`, `validatingWHConfig` and `clientConfig`

Reply:
`clientConfig` is not a `runtime.Object`. It's a field in both mutatinWebhookConfig and validatingWHConfig.